packages/grpc-js/src/transport.ts

```diff
@@ -101,28 +101,29 @@ class Http2Transport implements Transport {
   /**
    * The amount of time in between sending pings
    */
-  private keepaliveTimeMs = -1;
+  private readonly keepaliveTimeMs: number;
   /**
    * The amount of time to wait for an acknowledgement after sending a ping
    */
-  private keepaliveTimeoutMs: number = KEEPALIVE_TIMEOUT_MS;
+  private readonly keepaliveTimeoutMs: number;
   /**
-   * Timer reference for timeout that indicates when to send the next ping
+   * Indicates whether keepalive pings should be sent without any active calls
+   */
+  private readonly keepaliveWithoutCalls: boolean;
+  /**
+   * Timer reference indicating when to send the next ping or when the most recent ping will be considered lost.
    */
-  private keepaliveTimerId: NodeJS.Timeout | null = null;
+  private keepaliveTimeout: NodeJS.Timeout | null = null;
   /**
    * Indicates that the keepalive timer ran out while there were no active
    * calls, and a ping should be sent the next time a call starts.
    */
   private pendingSendKeepalivePing = false;
   /**
-   * Timer reference tracking when the most recent ping will be considered lost
-   */
-  private keepaliveTimeoutId: NodeJS.Timeout | null = null;
-  /**
-   * Indicates whether keepalive pings should be sent without any active calls
+   * Indicates when keepalives should no longer be performed for this transport. Used to prevent a race where a
+   * latent session.ping(..) callback is called after the transport has been notified to disconnect.
    */
-  private keepaliveWithoutCalls = false;
+  private keepaliveDisabled = false;
```

Comment:
I don't think we need a separate variable for this. We should just be able to check `session.destroyed`.

Reply:
The main consequence of that sequence of events is that one extra timer is started. And I don't think that's a significant enough problem that we need to track more state just to avoid it.

We could minimize the chances of that by explicitly calling `session.destroy()` at the end of the `setImmediate` block in `handleDisconnect`.